src/stirling/source_connectors/socket_tracer/mux_trace_bpf_test.cc

```diff
@@ -0,0 +1,158 @@
+
+    std::vector<std::string> tokens = absl::StrSplit(out, "\n");
+
+    int32_t client_pid;
+    if (!absl::SimpleAtoi(tokens[0], &client_pid)) {
+      return error::Internal("Could not extract PID.");
+    }
+
+    LOG(INFO) << absl::StrFormat("Client PID: %d", client_pid);
+
+    if (tokens[1] != thriftmux_client_output) {
+      return error::Internal(
+          absl::StrFormat("Expected output from thriftmux to be '%s', received '%s' instead",
+                          thriftmux_client_output, tokens[1]));
+    }
+    return client_pid;
+  }
+
+  ::px::stirling::testing::ThriftMuxContainer server_;
+};
+
+std::vector<mux::Record> ToRecordVector(const types::ColumnWrapperRecordBatch& rb,
+                                        const std::vector<size_t>& indices) {
+  std::vector<mux::Record> result;
+
+  for (const auto& idx : indices) {
+    mux::Record r;
+    uint24_t tag = static_cast<uint24_t>(rb[kMuxTagIdx]->Get<types::Int64Value>(idx).val);
+    r.req.tag = tag;
+    r.req.type = static_cast<int8_t>(rb[kMuxReqTypeIdx]->Get<types::Int64Value>(idx).val);
+    r.resp.tag = tag;
+    r.resp.type = static_cast<int8_t>(rb[kMuxRespTypeIdx]->Get<types::Int64Value>(idx).val);
+    result.push_back(r);
+  }
+  return result;
+}
+
+std::vector<mux::Record> GetTargetRecords(const types::ColumnWrapperRecordBatch& record_batch,
+                                          int32_t pid) {
+  std::vector<size_t> target_record_indices =
+      FindRecordIdxMatchesPID(record_batch, kMuxUPIDIdx, pid);
+  return ToRecordVector(record_batch, target_record_indices);
+}
+
+inline auto EqMux(const mux::Frame& x) {
+  return AllOf(Field(&mux::Frame::type, ::testing::Eq(x.type)),
+               Field(&mux::Frame::tag, ::testing::Eq(x.tag)));
+}
+
+inline auto EqMuxRecord(const mux::Record& x) {
+  return AllOf(Field(&mux::Record::req, EqMux(x.req)), Field(&mux::Record::resp, EqMux(x.resp)));
+}
+
+//-----------------------------------------------------------------------------
+// Test Scenarios
+//-----------------------------------------------------------------------------
+
+TEST_F(MuxTraceTest, Capture) {
+  StartTransferDataThread();
+
+  ASSERT_OK(RunThriftMuxClient());
+
+  StopTransferDataThread();
+
+  // Grab the data from Stirling.
+  std::vector<TaggedRecordBatch> tablets = ConsumeRecords(SocketTraceConnector::kMuxTableNum);
+  ASSERT_FALSE(tablets.empty());
+  types::ColumnWrapperRecordBatch record_batch = tablets[0].records;
+
+  std::vector<mux::Record> server_records = GetTargetRecords(record_batch, server_.process_pid());
+
+  mux::Record expected = {};
+  expected.req.tag = 1;
+  expected.req.type = 127;
+  expected.resp.tag = 1;
+  expected.resp.type = 127;
+
+  // TODO(ddelnano): Figure out why only the Rerr message is recorded.
+  // There should be a Tinit/Rinit, Tdispatch/Rdispatch and Tping/Rping as well
+  EXPECT_THAT(server_records, Contains(EqMuxRecord(expected)));
```

Comment:
@oazizi000 I'm only seeing a fraction of this data show up. Any tips of how to debug this? When I set a breakpoint in [AppendMessage](https://github.com/pixie-io/pixie/pull/351/files#diff-f870d234109a188f71ff069fa0ab63e4dda6ea1026a0e5bd6368c059f0d4280fR832) I only see it fire once. I see that [this](https://github.com/pixie-io/pixie/blob/2de76f03bcf34179edbcdf11780f94dca59886ce/src/stirling/source_connectors/socket_tracer/socket_trace_connector.cc#L987) logs a significant number of kProtocolMux packets but I see that they aren't always when the tracker is in a transferring state.

Reply:
The first thing to figure out is whether the issue is on the BPF side or in the user-space side. One thing you can do is to enable the `CONN_TRACE` for ConnTracker's that have the Mux protocol. You can do that with a call to `SetDebugTrace(2)` on the ConnTracker. That'll dump out a bunch of detailed information, but you can use that to figure out whether the expected events are arriving from BPF.

Reply:
Awesome, thanks for the tip. I'll give that a try.

Reply:
I was able to verify that all of the data is being captured through BPF and which mux messages are making it through to the stitcher code with the following steps:

```
# Update socket tracer to log all CONN_TRACE events
ddelnano@ddelnano-desktop:~/code/pixie$ git diff
diff --git a/src/stirling/source_connectors/socket_tracer/conn_tracker.h b/src/stirling/source_connectors/socket_tracer/conn_tracker.h
index d4aa4f749..a5ccbd7c4 100644
--- a/src/stirling/source_connectors/socket_tracer/conn_tracker.h
+++ b/src/stirling/source_connectors/socket_tracer/conn_tracker.h
@@ -619,7 +619,7 @@ class ConnTracker : NotCopyMoveable {
     stats_.Increment(StatKey::kValidRecords, result.records.size());
   }

-  int debug_trace_level_ = 0;
+  int debug_trace_level_ = 2;

# Run mux test with debug output and save to a file
root@ubuntu2004:/vagrant# ./bazel-bin/src/stirling/source_connectors/socket_tracer/mux_trace_bpf_test 2>&1  | tee output.txt

# Verify that the missing Tdispatch, Rdispatch messages are captured by BPF 
ddelnano@ddelnano-desktop:~/code/pixie$ grep 'StringString\|Deadline' output.txt
I1129 21:40:54.212054 35623 conn_tracker.cc:134] conn_id=[pid=35515 start_time_ticks=1059133 fd=108 gen=10617112283527] state=kCollecting remote_addr=-:-1 role=kRoleServer protocol=kProtocolMux Data event received: attr:[[ts=10617915725426 conn_id=[pid=35515 start_time_ticks=1059133 fd=108 gen=10617112283527] protocol=kProtocolMux role=kRoleServer dir=kIngress ssl=false source_fn=kSyscallRead pos=73 size=204 buf_size=204]] msg_size:204 msg:[\x00\x00\x00\xC8\x02\x00\x00\x02\x00\x03\x00(com.twitter.finagle.tracing.TraceContext\x00 9 E\xF5ld\x86\xA19 E\xF5ld\x86\xA19 E\xF5ld\x86\xA1\x00\x00\x00\x00\x00\x00\x00\x00\x00\x1Bcom.twitter.finagle.Retries\x00\x04\x00\x00\x00\x00\x00\x1Ccom.twitter.finagle.Deadline\x00\x10\x16\xBC"G%@ \xC0\x7F\xFF\xFF\xFF\xFF\xFF\xFF\xFF\x00\x00\x00\x00\x80\x01\x00\x01\x00\x00\x00\x05query\x00\x00\x00\x00\x0B\x00\x01\x00\x00\x00\x06String\x00]
I1129 21:40:54.212172 35623 conn_tracker.cc:134] conn_id=[pid=35646 start_time_ticks=1061007 fd=104 gen=10617107829483] state=kCollecting remote_addr=-:-1 role=kRoleClient protocol=kProtocolMux Data event received: attr:[[ts=10617924385846 conn_id=[pid=35646 start_time_ticks=1061007 fd=104 gen=10617107829483] protocol=kProtocolMux role=kRoleClient dir=kEgress ssl=false source_fn=kSyscallWriteV pos=77 size=200 buf_size=200]] msg_size:200 msg:[\x02\x00\x00\x02\x00\x03\x00(com.twitter.finagle.tracing.TraceContext\x00 9 E\xF5ld\x86\xA19 E\xF5ld\x86\xA19 E\xF5ld\x86\xA1\x00\x00\x00\x00\x00\x00\x00\x00\x00\x1Bcom.twitter.finagle.Retries\x00\x04\x00\x00\x00\x00\x00\x1Ccom.twitter.finagle.Deadline\x00\x10\x16\xBC"G%@ \xC0\x7F\xFF\xFF\xFF\xFF\xFF\xFF\xFF\x00\x00\x00\x00\x80\x01\x00\x01\x00\x00\x00\x05query\x00\x00\x00\x00\x0B\x00\x01\x00\x00\x00\x06String\x00]
I1129 21:40:54.212291 35623 conn_tracker.cc:134] conn_id=[pid=35646 start_time_ticks=1061007 fd=104 gen=10617107829483] state=kCollecting remote_addr=-:-1 role=kRoleClient protocol=kProtocolMux Data event received: attr:[[ts=10618450851988 conn_id=[pid=35646 start_time_ticks=1061007 fd=104 gen=10617107829483] protocol=kProtocolMux role=kRoleClient dir=kIngress ssl=false source_fn=kSyscallRead pos=73 size=48 buf_size=48]] msg_size:48 msg:[\x00\x00\x00,\xFE\x00\x00\x02\x00\x00\x00\x80\x01\x00\x02\x00\x00\x00\x05query\x00\x00\x00\x00\x0B\x00\x00\x00\x00\x00\x0CStringString\x00]
I1129 21:40:54.212399 35623 conn_tracker.cc:134] conn_id=[pid=35515 start_time_ticks=1059133 fd=108 gen=10617112283527] state=kCollecting remote_addr=-:-1 role=kRoleServer protocol=kProtocolMux Data event received: attr:[[ts=10618473421760 conn_id=[pid=35515 start_time_ticks=1059133 fd=108 gen=10617112283527] protocol=kProtocolMux role=kRoleServer dir=kEgress ssl=false source_fn=kSyscallWriteV pos=77 size=44 buf_size=44]] msg_size:44 msg:[\xFE\x00\x00\x02\x00\x00\x00\x80\x01\x00\x02\x00\x00\x00\x05query\x00\x00\x00\x00\x0B\x00\x00\x00\x00\x00\x0CStringString\x00]
StringString
```
I added more log statements to the mux stitcher code and verified that the following mux messages are passed to the protocol parsing code:
- client Rerr
- server Rerr
- client Tinit

What is missing is:
- server Rinit (response to Tinit)
- client / server T/Rping (these may or may not be present depending on how quickly finagle tries to healthcheck)
- client Tdispatch
- server Rdispatch

Everything except the Tping/Rping has an associated "Data even received" (as seen above) so it seems to rule out that the BPF side is the problem? However, I'm not sure what other layers exist between BPF identifying this data correctly and it being passed to the mux protocol code.

Any ideas on where to look next?